sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/nullExpressions.scala

```diff
@@ -128,6 +128,143 @@ case class IsNaN(child: Expression) extends UnaryExpression
 }
 
 /**
+ * An Expression accepts two parameters and returns null if both parameters are equal.
+ * If they are not equal, the first parameter value is returned.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a,b) - Returns null if a equals to b, or a otherwise.")
+case class NullIf(left: Expression, right: Expression) extends BinaryExpression {
+  override def nullable: Boolean = true
+  override def dataType: DataType = left.dataType
+
+  override def eval(input: InternalRow): Any = {
+    val valueLeft = left.eval(input)
+    val valueRight = right.eval(input)
+    if (valueLeft.equals(valueRight)) {
+      null
+    } else {
+      valueLeft
+    }
+  }
+
+  override def genCode(ctx: CodegenContext, ev: ExprCode): String = {
+    val leftGen = left.gen(ctx)
+    val rightGen = right.gen(ctx)
+    s"""
+      ${leftGen.code}
+      ${rightGen.code}
+      boolean ${ev.isNull} = false;
+      ${ctx.javaType(dataType)} ${ev.value} = ${ctx.defaultValue(dataType)};
+      if (${ctx.genEqual(dataType, leftGen.value, rightGen.value)}) {
+        ${ev.isNull} = true;
+      } else {
+        ${ev.value} = ${leftGen.value};
+      }
+    """
+  }
+}
+
+/**
+ * An Expression accepts two parameters and returns the second parameter if the value
+ * in the first parameter is null; if the first parameter is any value other than null,
+ * it is returned unchanged.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a,b) - Returns b if a is null, or a otherwise.")
+case class Nvl(left: Expression, right: Expression) extends BinaryExpression {
```

Comment:
I will say, yes, kind of. Here is what I found: [difference](http://stackoverflow.com/questions/950084/oracle-differences-between-nvl-and-coalesce)